### DISCUSSION TITLE
6-Thunder-Loan-Audit make/forge build errors

### DISCUSSION THREAD

**Discussion by Trisial** [Cyfrin/security-and-auditing-full-course-s23#266] [Q&A] [answered]
Hi everyone,
I have a problem while cloning the 6-Thunder-Loan-Audit repo into foundry. after cloning and running the `make` command i get an error regarding incorrect solidity versions:

```
[⠃] Compiling...2024-12-20T17:29:17.661219Z ERROR foundry_compilers::resolver: failed to resolve versions
[⠊] Compiling...
Error: Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
Encountered invalid solc version in lib/openzeppelin-contracts/contracts/proxy/ERC1967/ERC1967Utils.sol: No solc version exists that matches the version requirement: ^0.8.21
make: *** [build] Error 1
```

It creates a problem, because all the files and my solc is set to 0.8.20. 

I have tried changing my solc to 0.8.21 or changing each file to ^0.8.20 however it also created more errors. 

This is the errors i got while changing the versions:
```
[⠊] Compiling...
[⠊] Compiling 66 files with Solc 0.8.21
[⠒] Solc 0.8.21 finished in 953.38ms
Error: Compiler run failed:
Error: Internal exception in StandardCompiler::compile: /solidity/libsolidity/interface/Natspec.cpp(89): Throw in function static Json::Value solidity::frontend::Natspec::userDocumentation(const solidity::frontend::ContractDefinition &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
Internal exception in StandardCompiler::compile: /solidity/libsolidity/interface/Natspec.cpp(89): Throw in function static Json::Value solidity::frontend::Natspec::userDocumentation(const solidity::frontend::ContractDefinition &)
Dynamic exception type: boost::wrapexcept<solidity::langutil::InternalCompilerError>
std::exception::what: Solidity assertion failed
[solidity::util::tag_comment*] = Solidity assertion failed
```

I have ran out of ideas how to repair it and it unables me to continue with the course as i cannot `forge build` or run slither etc.
I have also tried running foundryup beforehand and reinstalling it, because I saw some similar problems in some other discussions.

Do you have any idea how to solve it?

**Comment by EngrPips** [comment]
Hello @Trisial, I haven't encountered this error before so it is tough for me to figure out what could be wrong but I asked Gemini Ai and below is what it suggested.

The error message you're encountering suggests an issue with the Solidity compiler's internal logic, specifically related to how it handles Natspec documentation. 

**Natspec** is a documentation format used in Solidity to describe the purpose, parameters, and return values of contracts and functions. 

**Here's a breakdown of the error and potential solutions:**

* **Internal Compiler Error:** 
    - This indicates a bug within the Solidity compiler itself. These errors are usually rare and often related to specific compiler versions or edge cases in the Solidity language.

* **Natspec Parsing Issue:** 
    - The error likely stems from the compiler's inability to correctly parse or process the Natspec comments within your Solidity code.

**Troubleshooting Steps:**

1. **Update Solidity Compiler:**
   - **Check for Updates:** Update to the latest stable version of the Solidity compiler. Newer versions often include bug fixes and improvements that may resolve this issue. You can typically update using a package manager like npm (if you're using a tool like Truffle or Hardhat) or by downloading the latest compiler binaries.

2. **Review Natspec Comments:**
   - **Syntax and Formatting:** Carefully review the Natspec comments in your code. Ensure they adhere to the correct syntax and formatting conventions. 
     - Use the correct tags (e.g., `@param`, `@return`, `@notice`)
     - Maintain proper indentation and spacing.
     - Avoid any unusual or unsupported characters.
   - **Minimal Example:** Create a minimal example that reproduces the error. This will help you isolate the problematic code and make it easier to debug.

3. **Clear Compiler Cache:**
   - If you're using a tool like Truffle or Hardhat, clear the compiler cache. This can sometimes resolve issues caused by cached compilation artifacts.

4. **Check for Compiler Bugs:**
   - If you've tried the above steps and are still encountering the error, it might be a compiler bug. 
     - **Report the Issue:** If you suspect a compiler bug, report it to the Solidity team on their GitHub repository. Include the following:
        - The specific version of the Solidity compiler you're using.
        - A minimal reproducible example of the code that triggers the error.
        - The complete error message.

**Example of Incorrect Natspec:**

```solidity
contract MyContract {
    // Incorrect: Missing closing tag
    @notice This is a contract 

    function myFunction() public { 
        // ...
    }
}
```

**Example of Correct Natspec:**

```solidity
contract MyContract {
    /**
     * @notice This is a contract description.
     */
    function myFunction() public { 
        // ...
    }
}
```

I hope this helps! If you can provide the specific Solidity code snippet that's causing the error, I might be able to give you more tailored guidance.

**Reply by EngrPips** [reply]
Please look into this and let us know if it helps.

**Reply by Trisial** [reply, marked as answer]
Hi, thanks for the help. I tried your solution but unfortunately it did not work. However, I managed to solve the issue. I figured out that by cloning the repository it was installing a different version of openzeppelin-contracts than the repo used (v5.1.0), which is why it was not compiling as the newer version used solidity 0.8.21. So I installed the older version of openzeppelin v5.0.0 and it managed to compile.

**Reply by Trisial** [reply]
However I find this confusing that it worked for you and my friend with whom I checked that as well without any problems so I assume it was installing the right version of openzeppelin on your computers and not on mine.

**Reply by EngrPips** [reply]
Yeah, it makes sense that it worked for me because I took the course many months back. I am glad you figured it out anyways.
